File: Algorithms/Graph Theory/Java Implementations/GraphAdjacencyList.java
/**Generic Graph Data Structure using Adjacency List, implemented by Arpan Pathak **/
import java.util.*;
class Edge<T,K>
{
	public T v;   K w;
	public Edge(T vertex,K weight) { v=vertex; w=weight; }
}
class GraphList<T,V>
{
	Map<T, List< Edge<T,V>> > G;
	GraphList()
	{
		G=new HashMap<T, List<Edge<T,V>> >();
	}
	void addEdge(T u,T v,V weight)
	{
		if(!G.containsKey(u))
			G.put(u, new LinkedList<Edge<T,V>>() );
		G.get(u).add(0, new Edge<T,V>(v,weight));
	}
	void display()
	{
		for(T vertex: G.keySet())
		{
			System.out.print(vertex+"=>");
			for(Edge<T,V> iterator: G.get(vertex))
				System.out.print("("+iterator.v+","+iterator.w+"),");
			System.out.println();
		}
	}
}
public class GraphAdjacencyList {
	public static void main(String[] args) {
		GraphList<String,Integer> g=new GraphList<>();
		g.addEdge("Delhi", "Kolkata", 80);
		g.addEdge("Delhi", "Channai", 23);
		g.addEdge("Kolkata", "Delhi", 50);
		g.display();
	}
}

File: Algorithms/Graph Theory/C++ Implementations/GraphAdjacencyList.cpp
/** Directed Graph using adjacency list, implemented by Arpan Pathak **/
#include <bits/stdc++.h>
#define INF LONG_MAX
using namespace std;
template<class K,class V>
class Edge
{
        public:
            K vertex; V weight;
            Edge(K vertex,V weight): vertex(vertex),weight(weight) { }
};
template<class K,class V>
class Graph
{
    map<K,list<Edge<K,V>> > G;
    int n;

    public:
        Graph(int n): n(n) { }

        void addEdge(K u,K v,V w) { G[u].push_front(Edge<K,V>(v,w)); }
        // typename required for this kind of user defined type object iterator
         V getWeight(K u,K v) {
            for(typename list<Edge<K,V>>::iterator i=G[u].begin();i!=G[u].end();++i)
                if(i->vertex==v) return i->weight;
            return INF;
        }
        void print()
        {
            for(auto i: G)
            {
                cout<<i.first<<"==>";
                for(typename list<Edge<K,V>>::iterator j=G[i.first].begin(); j!=G[i.first].end(); ++j)
                    cout<<"("<<j->vertex<<","<<j->weight<<"), ";
                cout<<endl;
            }
        }
};
int main()
{
    Graph<string,int> G(5);
    G.addEdge("Delhi","Mumbai",15);
    G.addEdge("Delhi","Kolkata",20);
    G.addEdge("Dhaka","Kolkata",50);
    G.print();
}

